determined-ai/determined
# feat: cluster wide startup hook for tasks

## Ticket

## TODO
- [x] slurm test
- [x] docs and release notes
- [x] an e2e tests that shows the hook running. via logs? or command
## Description


- (sep from this PR) should we be merging RM TCDs with clusterwide TCDs instead of overriding them?
https://github.com/determined-ai/determined/blob/d5f807da0ecded7d7b4c6afbf37bbfce8d0233f2/master/internal/rm/kubernetesrm/kubernetes_resource_manager.go#L631

## Test Plan


- set startup hook various task container defaults: master config, resource manager
  - check that RM startuphook override cluster wide one
- unit tests

basic example
```
        task_container_defaults:
          startup_hook: |
            echo "TCD startup thing from master config"
            echo "task type is $DET_TASK_TYPE"
```

```
hbp ➜ d/determined-ee git:(hz-ee-startuphook) ✗ det cmd run 'ls $HOME; date; ls /run/determined/'
Launched command (id: 3d1d922b-4edd-4556-b2df-a94806138a4a).
[2024-04-09T21:50:07.432659Z]          || INFO: Scheduling Command (singularly-enormous-ibex) (id: 3d1d922b-4edd-4556-b2df-a94806138a4a.1)
[2024-04-09T21:50:11.340487Z]          || INFO: HPC Job ID: 3
[2024-04-09T21:50:11.342877Z]          || INFO: Command (singularly-enormous-ibex) was assigned to an agent
[2024-04-09T21:50:16.739730Z] [cont=0] || WARNING: [10466] determined: falling back to naive proxy ip resolution (error=None)
[2024-04-09T21:50:17.238141Z] [cont=0] || + test -f /run/determined/dynamic-tcd-startup-hook.sh
[2024-04-09T21:50:17.238237Z] [cont=0] || + source /run/determined/dynamic-tcd-startup-hook.sh
[2024-04-09T21:50:17.238266Z] [cont=0] || ++ echo hi from master tcd slurm
[2024-04-09T21:50:17.238279Z] [cont=0] || + set +x
[2024-04-09T21:50:17.238379Z] [cont=0] || hi from master tcd slurm
[2024-04-09T21:50:17.246344Z] [cont=0] || Tue Apr  9 21:50:17 UTC 2024
[2024-04-09T21:50:17.248080Z] [cont=0] || command-entrypoint.sh
[2024-04-09T21:50:17.248151Z] [cont=0] || dynamic-tcd-startup-hook.sh
[2024-04-09T21:50:17.248172Z] [cont=0] || etc
[2024-04-09T21:50:17.248190Z] [cont=0] || info
[2024-04-09T21:50:17.248208Z] [cont=0] || pythonuserbase
[2024-04-09T21:50:17.248223Z] [cont=0] || ship-logs.sh
[2024-04-09T21:50:17.248234Z] [cont=0] || ship_logs.py
[2024-04-09T21:50:17.248260Z] [cont=0] || singularity-entrypoint-wrapper.sh
[2024-04-09T21:50:17.248267Z] [cont=0] || task-setup.sh
[2024-04-09T21:50:17.248273Z] [cont=0] || train
[2024-04-09T21:50:17.248279Z] [cont=0] || workdir
[2024-04-09T21:50:21.314010Z]          || INFO: resources exited successfully with a zero exit code
[2024-04-09T21:50:21.330339Z]          || INFO: Command (singularly-enormous-ibex) was terminated: allocation stopped early after all resources exited: resources exited successfully with a zero exit code
Task log stream ended. To reopen log stream, run: det task logs -f 3d1d922b-4edd-4556-b2df-a94806138a4a
```

## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.




[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

https://hpe-aiatscale.atlassian.net/browse/RM-105
